server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4307,6 +4307,31 @@ public void afterRefresh(boolean didRefresh) {
         }
     }
 
+    /**
+     * Reset the current engine to a new one.
+     *
+     * Calls {@link Engine#prepareForEngineReset()} on the current engine, then closes it, and loads a new engine without
+     * doing any translog recovery.
+     */
+    public void resetEngine() {
+        assert Thread.holdsLock(mutex) == false : "resetting engine under mutex";
```

Comment:
Can we assert that all the permits are held when this is called? also, I think that we should leave a TODO here to make stronger assertions about the expected state before this is called

Reply:
Sounds good